topics/sw_concepts/sw_concept_slides.md

```diff
@@ -0,0 +1,439 @@
+Auch ist es jetzt einfacher, neue Regeln hinzuzufügen.
+<https://github.com/soco21/soco21-group8/tree/fb763220b26d3ed0260580d944b92fe9172f6c7b/assignment-1/exercise3/src/main/java/ch/uzh/group8/assignment1/exercise3/movevalidator>
+
+```python
+class StartPieceValid:
+class TargetFieldEmpty:
+class MoveLength:
+class MoveIsForwardIfNotKing:
+class MoveIsDiagonal:
+class OpponentPieceBetweenJump:
+class NoOtherJumpMovePossible:
+```
+
+Open Closed Beispiel 3
+-------
+
+Und es ermöglicht composition:
+
+```python
+  moveValidators = [
+            startPieceValid,
+            targetFieldEmpty,
+            moveLength,
+            moveIsForwardIfNotKing,
+            moveIsDiagonal,
+            opponentPieceBetweenJump,
+            noOtherMoveToJumpPossible
+  ]
+  winCondition = new WinCondition([
+            startPieceValid,
+            targetFieldEmpty,
+            moveLength,
+            moveIsForwardIfNotKing,
+            moveIsDiagonal,
+            opponentPieceBetweenJump
+  ])
+```
+
+Liskovs substitution principle
+-------
+
+* Implementationen eines Interfaces müssen ausgetauscht werden können.
+* Sie nehmen mindestens die gleichen Input werte an (dürfen auch mehr).
+* Sie geben maximal die gleichen Rückgabewerte zurück.
+* Werfen maximal die gleichen Exceptions (oder Subtypen der Exceptions).
+
+Beispiel Liskovs substitution principle Verletzung
+-------
+
+```python
+class Vehicle:
+    pass
+
+class Car(Vehicle):
+    pass
+
+class Bicycle(Vehicle):
+    pass
+
+class Garage:
+    def park(self, vehicle: Vehicle):
+        pass
+
+class BicycleGarage(Garage):
+    def park(self, vehicle: Vehicle):
+        if (not isinstance(vehicle, Bicycle)):
+            raise ValueError("BicycleGarage can only park bicycles")
+
+```
+
+Interface Segregation
+-------
+
+* Um die Kopplung zu reduzieren, kann man auch Interfaces aufteilen.
+* Eine Klasse kann mehrere Interfaces implementieren
+* Man will mit einer Klasse nicht benötigte Methoden nicht implementieren
+* Man will nicht, dass jemand plötzlich eine Methode nutzt, die man gar nicht zur Verfügung stellen wollte.
+
+Interface Segregation Example
+------
+
+![Interface Segregation](images/interface-segregation/java-interface-segregation.png){height=95%}
+
+Dependency Inversion
+------
+
+* Um 2 Klassen zu entkoppeln kann ein Interface dazwischen geschalten werden.
+* Das verhindert, dass eine High Level Komponente von einer Low Level Komponente abhängt.
+
+Dependency Inversion Beispiel 1
+-----
+
+![No dependency inversion](images/dependency-inversion/no-dependency-inversion.png)
+
+Dependency Inversion Beispiel 2
+-----
+
+![Dependency inversion](images/dependency-inversion/with-dependency-inversion.png){height=95%}
+
+Design Patterns
+-------
+
+<https://github.com/iluwatar/java-design-patterns>
+
+Factory (method)
+-------
+
+<https://github.com/iluwatar/java-design-patterns/blob/master/factory-method/README.md>
```

Comment:
Diese freien Links sollten wenigstens einen sinnvollen Namen haben. Damit weiss, warum der Link hier steht. Grundsätzlich sollten die Folien aber selbsterklärend sein und Links nur als Ergänzung dienen.

Reply:
Werde ich in weiteren PR's angehen